crates/router/src/core/routing/helpers.rs

```diff
@@ -543,3 +554,334 @@ pub fn get_default_config_key(
+        .parse_value::<routing_types::DynamicRoutingAlgorithmRef>("DynamicRoutingAlgorithmRef")
+        .change_context(errors::ApiErrorResponse::InternalServerError)
+        .attach_printable("unable to parse dynamic_algorithm_ref")?;
+
+    let success_based_routing_id = dynamic_routing_algorithm_ref
+        .success_based_algorithm
+        .ok_or(errors::ApiErrorResponse::GenericNotFoundError {
+            message: "unable derive success_based_algorithm from dynamic algorithm ref".to_string(),
+        })?
+        .algorithm_id
+        .ok_or(errors::ApiErrorResponse::GenericNotFoundError {
+            message: "unable to find algorithm id in success based algorithm".to_string(),
+        })?;
+
+    let key = format!(
+        "{}_{}",
+        business_profile.get_id().get_string_repr(),
+        success_based_routing_id.get_string_repr()
+    );
+
+    if let Some(config) = get_cached_success_routing_config_for_profile(state, key.as_str()).await {
+        Ok(config.as_ref().clone())
+    } else {
+        let success_rate_algorithm = state
+            .store
+            .find_routing_algorithm_by_profile_id_algorithm_id(
+                business_profile.get_id(),
+                &success_based_routing_id,
+            )
+            .await
+            .change_context(errors::ApiErrorResponse::ResourceIdNotFound)
+            .attach_printable("unable to retrieve success_rate_algorithm for profile")?;
+
+        let success_rate_config = success_rate_algorithm
+            .algorithm_data
+            .parse_value::<routing_types::SuccessBasedRoutingConfig>("SuccessBasedRoutingConfig")
+            .change_context(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable("unable to parse success_based_routing_config struct")?;
+
+        refresh_success_based_routing_cache(state, key.as_str(), success_rate_config.clone()).await;
+
+        Ok(success_rate_config)
+    }
+}
+
+/// metrics for success based dynamic routing
+#[cfg(all(feature = "v1", feature = "dynamic_routing"))]
+pub async fn metrics_for_success_based_routing(
+    state: &SessionState,
+    key_store: &domain::MerchantKeyStore,
+    payment_attempt: &storage::PaymentAttempt,
+    routable_connectors: Vec<routing_types::RoutableConnectorChoice>,
+) -> RouterResult<()> {
+    let client = state
+        .grpc_client
+        .dynamic_routing
+        .success_rate_client
+        .as_ref()
+        .ok_or(errors::ApiErrorResponse::GenericNotFoundError {
+            message: "success_rate gRPC client not found".to_string(),
+        })?;
+
+    let payment_connector = &payment_attempt.connector.clone().ok_or(
+        errors::ApiErrorResponse::GenericNotFoundError {
+            message: "unable to derive payment connector from payment attempt".to_string(),
+        },
+    )?;
+
+    let key_manager_state = &state.into();
+
+    let business_profile = state
+        .store
+        .find_business_profile_by_profile_id(
```

Comment:
Is this call needed here? Since this method is called in a payments flow, business profile should already be fetched?